api/src/routes/donate.ts

```diff
@@ -214,3 +201,164 @@ export const donateRoutes: FastifyPluginCallbackTypebox = (
+        });
+
+        const stripeSubscription = await stripe.subscriptions.create({
+          customer: stripeCustomer.id,
+          items: [
+            {
+              plan: `${donationSubscriptionConfig.duration[duration]}-donation-${amount}`
+            }
+          ],
+          payment_behavior: 'default_incomplete',
+          payment_settings: { save_default_payment_method: 'on_subscription' },
+          expand: ['latest_invoice.payment_intent']
+        });
+
+        if (
+          stripeSubscription.latest_invoice &&
+          typeof stripeSubscription.latest_invoice !== 'string' &&
+          stripeSubscription.latest_invoice.payment_intent &&
+          typeof stripeSubscription.latest_invoice.payment_intent !==
+            'string' &&
+          stripeSubscription.latest_invoice.payment_intent.client_secret !==
+            null
+        ) {
+          const clientSecret =
+            stripeSubscription.latest_invoice.payment_intent.client_secret;
+          return reply.send({
+            subscriptionId: stripeSubscription.id,
+            clientSecret
+          });
+        } else {
+          throw new Error('Stripe payment intent client secret is missing');
+        }
+      } catch (error) {
+        fastify.log.error(error);
+        void reply.code(500);
+        return reply.send({
+          error: 'Donation failed due to a server error.'
+        });
+      }
+    }
+  );
+
+  fastify.post(
+    '/donate/charge-stripe',
+    {
+      schema: schemas.chargeStripe
+    },
+    async (req, reply) => {
+      try {
+        const { email, amount, duration, subscriptionId } = req.body;
+
+        const subscription =
+          await stripe.subscriptions.retrieve(subscriptionId);
+        const isSubscriptionActive = subscription.status === 'active';
+        const productId =
+          subscription.items?.data[0]?.plan?.product?.toString();
+        const isStartedRecently = inLastFiveMinutes(
+          subscription.current_period_start
+        );
+        const isProductIdValid =
+          typeof productId == 'string' &&
+          allStripeProductIdsArray.includes(productId);
+
+        const isValidCustomer = typeof subscription.customer === 'string';
+        if (
+          isSubscriptionActive &&
+          isProductIdValid &&
+          isStartedRecently &&
+          isValidCustomer
+        ) {
+          // TODO(Post-MVP) new users should not be created if user is not found
+          const user = await findOrCreateUser(fastify, email);
+          const donation = {
+            userId: user.id,
+            email,
+            amount,
+            duration,
+            provider: 'stripe',
+            subscriptionId,
+            customerId: subscription.customer as string,
+            // TODO(Post-MVP) migrate to startDate: new Date()
+            startDate: {
+              date: new Date().toISOString(),
+              when: new Date().toISOString().replace(/.$/, '+00:00')
+            }
```

Comment:
Just a query, but is this replicating some loopback magic? I can't see where this came from.

Reply:
This is how the startDate is recorded in the DB because of some loopback magic. Once we are off loopack, we should normalize the dates across.

Reply:
Okay, thanks for confirming.